### DISCUSSION TITLE
How to add custom arguments to a client extension query

### DISCUSSION THREAD

**Discussion by galenwinsor** [prisma/prisma#25108] [Q&A]
### Question

I'm using client extensions for four tables in my database to create a `Notification` record every time `update`, `create`, etc. gets called on one of the four tables. `Notification` records include information about the user who applied the change, and I'm stuck figuring out how to pass the current user's UID to the `update`, `create`, etc. query.

We've tried/thought of the following solutions:
1. Add an additional parameter to the `args` object that's passed to the query.
2. Create a custom query, e.g., `updateWithNotification`.
3. Use `@ts-ignore` and just trust that the caller passed `currentUserUID`...

Option 2 would involve a large-scale refactor, and unclear API expectations for future developers. Option 3 is obviously bad. I've tried Option 1 a few different ways using Prisma helpers, but keep getting TypeScript errors.

Is it possible to modify the type of `args` that's passed to a builtin query in a client extension? Or is this not supported?

If folks have other ideas for how to approach this, that would also be appreciated!

We'd like to do something like this: 

```typescript
export default Prisma.defineExtension((client) => {
	query: {
		study: {
			async update<A>({ args, query }: {
				args: Prisma.Exact<A, Prisma.Args<Study, "update"> & { currentUserUID: string }>;
				query: // unsure what this type would be
			}) {
				const { where, data, currentUserUID } = args;

				// ... logic using data and where
			
				await client.notification.create({ ... });
				
				return await query(...);
			}
		}
	},
	// ... rest of extension
}
```

#### Notifications
```prisma
model Notification {
  uid         String      @id @default(uuid())
  studyUID    String
  userUID     String
  time        DateTime    @default(now())
  action      Action
  itemChanged ItemChanged

  fileUID       String?
  trialUID      String?
  individualUID String?

  //Relations 1-1 
  file       File?       @relation(fields: [fileUID], references: [uid], onUpdate: Cascade, onDelete: Cascade)
  individual Individual? @relation(fields: [individualUID], references: [uid], onDelete: Cascade, onUpdate: Cascade)
  trial      Trial?      @relation(fields: [trialUID], references: [uid], onDelete: Cascade, onUpdate: Cascade)
  study      Study       @relation(fields: [studyUID], references: [uid], onDelete: Cascade, onUpdate: Cascade)
}
```

#### Study (one of the four tables that triggers notifications)

```prisma
model Study {
  uid         String    @id @default(uuid())
  createdAt   DateTime  @default(now())
  description String?   @db.VarChar(250) // Short (<250 characters) description of the study
  url         String?   @unique
  endDate     DateTime? @db.DateTime
  leaderUID   String? // Main lead on the study
  name        String
  overview    String?   @default("# Study Funding \n # Publications") @db.VarChar(2000) // HTML for the overview page
  ownerUID    String // Owner of the data on a study
  updatedAt   DateTime  @updatedAt
  startDate   DateTime  @default(now())
  archived    Boolean   @default(false)

  // Relations (1-1)
  thumbnail StudyThumbnail?

  // Relations (1-n)
  leader User? @relation("leader", fields: [leaderUID], references: [uid])
  owner  User  @relation("owner", fields: [ownerUID], references: [uid])

  // Relations (m-n)
  userStudyRoles      UserStudyRole[]
  metadataFieldValues MetadataFieldValue[]

  // Relations (1-n)
  individuals    Individual[]
  trials         Trial[]
  metadataFields MetadataField[]
  files          StudyFile[]
  notifications  Notification[]

  @@index([uid, name(sort: Desc)]) // Sort on name field by default
}
```

- OS: macOS
- Database: PostgreSQL
- Node.js version: 20.9

- Prisma: 5.18

**Comment by ludralph** [comment]
Hi @galenwinsor 👋 

Would using a helper function to modify the args object before passing it to the query work for you. For example:

```ts
function addCurrentUserUID<A>(args: Prisma.Exact<A, Prisma.Args<Study, "update"> & { currentUserUID: string }>) {
  const { currentUserUID, ...rest } = args;
  return {
    ...rest,
    data: {
      ...rest.data,
      userUID: currentUserUID,
    },
  };
}

export default Prisma.defineExtension((client) => ({
  query: {
    study: {
      async update({ args, query }) {
        const modifiedArgs = addCurrentUserUID(args);
        await client.notification.create({
          data: {
            studyUID: modifiedArgs.where.id,
            userUID: modifiedArgs.data.userUID,
            action: 'UPDATE',
            itemChanged: 'Study',
          },
        });
        return await query(modifiedArgs);
      },
    },
  },
}));
```
The `addCurrentUserUID` function modifies the `args` object to include the `currentUserUID` in the `data` field. This modified args object is then used in the query.

**Reply by galenwinsor** [reply]
Hi @ludralph! Thanks for the reply. I think I'd still get TypeScript errors, since the type of `args` still doesn't include `currentUserUID`. I don't actually want to pass `currentUserUID` to the query at all, I just want to be able to use it in creating the notification.

**Comment by galenwinsor** [comment]
Hi @ludralph, bumping this ^

**Comment by ludralph** [comment]
Hi 

One approach that could work for your use case is to use the `$allOperations` method in your client extension. This method allows you to modify all query methods present in Prisma Client, including the ones you're interested in (update, create, etc.). 
```ts
const prisma = new PrismaClient().$extends({
  query: {
    $allOperations({ model, operation, args, query }) {
      // Check if the model is one of the four you're interested in
      if (['Study', 'File', 'Trial', 'Individual'].includes(model)) {
        // Extract the currentUserUID from args
        const { currentUserUID, ...restArgs } = args;

        // Perform the original query
        const result = query(restArgs);

        // Create a notification
        if (currentUserUID) {
          prisma.notification.create({
            data: {
              userUID: currentUserUID,
              studyUID: result.uid, // Assuming all models have a uid field
              action: operation,
              itemChanged: model,
              // ... other fields as needed
            }
          });
        }

        return result;
      }

      // For other models, just perform the original query
      return query(args);
    },
  },
});
```
This approach allows you to pass the `currentUserUID` as part of the args for any operation on the models you're interested in. It will be extracted and used to create the Notification, while the rest of the args are passed to the original query.
To use this, you would call your Prisma methods like this:
```ts
await prisma.study.update({
  where: { uid: 'some-uid' },
  data: { /* your update data */ },
  currentUserUID: 'current-user-uid'
});
```
However, there's a caveat: this approach doesn't provide type safety for the `currentUserUID` parameter out of the box. You would need to extend the types for each model's operations to include this additional parameter.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by IanvanZyl636** [comment]
@galenwinsor did you ever manage to find a solution to add a custom property to all models and all operators?
